[DQL-204] Enhancements to the data request functionality

MarkCalvert opened the pull request:
This PR contains changes for the following stories:

[[DQL2-31]](https://salsadigital.atlassian.net/browse/DQL2-31)Close data request Circumstances enhancements

- Add template helper get_closing_circumstances
- Added closing circumstances dropdown to 'close_datarequest_form.html' with JS logic to show/hide condition elements
- Added conditional logic for througout extension for 'closing_circumstances_enabled'
- Updated unit tests to default to using 'tk.h.closing_circumstances_enabled = False'

[[DQL2-32]](https://salsadigital.atlassian.net/browse/DQL2-32) Configurable circumstances list
- Added new admin config 'datarequest_circumstances'

[[DQL2-33]](https://salsadigital.atlassian.net/browse/DQL2-33) Close data request finalisation
- Added new columns to datareqest table: close_circumstance & approx_publishing_date
- Added close_circumstance & approx_publishing_date values to action 'close_datarequest'
- Show close_circumstance & approx_publishing_date values on 'additional_info.html'

[[DQL2-40]](https://salsadigital.atlassian.net/browse/DQL2-40) Automated testing considerations
ded BDD tests for datarequest circumstances
- Added CKAN config 'ckan.datarequests.enable_closing_circumstances = True' to test.ini
- Updated 'create-test-data.sh' to add new admin config value for 'ckan.datarequests.closing_circumstances'
-  Updated 'create-test-data.sh' to create test data datasets
- Added new BDD steps 'I log in and go to admin config page',  'I go to admin config page', 'I log in and create a datarequest'
- Updated 'datarequest.feature' to use new step 'I log in and create a datarequest

This new feature requires 1 new CKAN config file update:
Enable or disable circumstances for closing data requests. Default value is False
ckan.datarequests.enable_closing_circumstances = True

This new feature requires to set a new CKAN sysadmin config value (/ckan-admin/config) for
'Data request closing circumstances':
Released as open data|nominate_dataset
Open dataset already exists|nominate_dataset
Partially released|nominate_dataset
To be released as open data at a later date|nominate_approximate_date
Data openly available elsewhere
Not suitable for release as open data
Requested data not available/cannot be compiled
Requestor initiated closure


@duttonw & @ThrawnCA for your review.

Commit:
[DQL2-32]
Added new admin config 'datarequest_circumstances'
[DQL2-33]
Added new columns to datareqest table:
- close_circumstance
- approx_publishing_date
Added close_circumstance & approx_publishing_date values to action 'close_datarequest'
Show close_circumstance & approx_publishing_date values on 'additional_info.html'
[DQL2-31]
- Add template helper get_closing_circumstances
- Added closing circumstances dropdown to 'close_datarequest_form.html' with JS logic to show/hide condition elements
- Refactored '_dictize_datarequest' & '_undictize_datarequest_basic' methods to the 'db.py' model
- Added conditional logic for throuhout extension for 'closing_circumstances_enabled'

Commit:
[DQL2-31]
- Updated close accepted_dataset dropdown to use dataset id instead of name as its value
- Updated validate_datarequest_closing to use 'package_id_exists' to validate accepted_dataset_id

Commit:
[DQL2-40]
- Added BDD tests for datarequest circumstances
- Added CKAN config 'ckan.datarequests.enable_closing_circumstances = True' to test.ini
- Updated 'create-test-data.sh' to add new admin config value for 'ckan.datarequests.closing_circumstances'
-  Updated 'create-test-data.sh' to create test data datasets
- Added new BDD steps 'I log in and go to admin config page',  'I go to admin config page', 'I log in and create a datarequest'
- Updated 'datarequest.feature' to use new step 'I log in and create a datarequest

[DQL2-33]
- Updated validation error messages to Sentence casing
- Update additinal_info to conditionally show 'approx_publishing_date' or 'accepted_dataset' rows if the values exist

Commit:
Allow unit tests failures in circleci

Commit:
[DQL2-33]
- Updated approx_publishing_date to use 'render_datetime 'helper function timestamp string as a localised date

Commit:
[DQL2-31]
- Reverted changes back to original way of using '_dictize_datarequest' & '_undictize_datarequest_basic' and not using model relationships
- Re-enabled unit test failures
- Updated unit tests to default to using 'tk.h.closing_circumstances_enabled = False'

Commit:
[DQL2-31]
Increase random number range to descrease change of duplicates

Commit:
[DQL-31]
- Updated CLOSE_CIRCUMSTANCE_MAX_LENGTH from 100 to 255

Commit:
[DQL-31]
- Updated variable name data_request to 'datarequest' to keep same standard in the file
- Updated _undictize_datarequest_closing_circumstances default values to None when empty string

salsa-nathan committed:
Merge pull request #1 from MarkCalvert/feature/Enhancements_to_the_data_request_functionality

DQL-204: Enhancements to the data request functionality

MarkCalvert commented on requirements-dev.txt:
```diff
@@ -8,7 +8,7 @@ nose_parameterized==0.3.3
 lxml==4.5.0
 splinter>=0.13.0
--e git+https://github.com/qld-gov-au/ckanext-data-qld@develop#egg=ckanext-data_qld
+-e git+https://github.com/salsadigitalauorg/ckanext-data-qld@feature/Enhancements_to_the_data_request_functionality#egg=ckanext-data_qld
```
Once data-qld PR (https://github.com/qld-gov-au/ckanext-data-qld/pull/71) has been approved and merged into develop, I will update this to point to https://github.com/qld-gov-au/ckanext-data-qld@develop

duttonw replied:
https://github.com/qld-gov-au/ckanext-data-qld/pull/71 has been merged (please make a note that we need to have this version and ckanext-data-qld (16 june 2020) tagged and released at the same time.

Commit:
Updated requirments-dev to use orginal repo from qld-gov-au

ThrawnCA commented on ckanext/datarequests/actions.py:
```diff
@@ -562,8 +573,9 @@ def close_datarequest(context, data_dict):
 
     data_req.closed = True
-    data_req.accepted_dataset_id = data_dict.get('accepted_dataset_id', None)
+    data_req.accepted_dataset_id = data_dict.get('accepted_dataset_id') or None
```
Won't this throw an exception if the key is not present? The previous syntax seems more robust.

duttonw replied:
since we want None anyway, this code just makes it more obvious, in the spec https://docs.python.org/2/library/stdtypes.html#dict.get if the second variable is not set, then None is the default so doing the or check would just set the same variable (from what the docs state) 
of course we could remove the ```or None``` and create a unit test to verify that it will works as expected  (or just rerun the current set)

MarkCalvert replied:
The `or` command uses boolean logic, so if `data_dict.get('accepted_dataset_id')` has no key present or the value is '' or None, the boolean value will be false and the value to the right of the 'or' command will be used. https://realpython.com/python-or-operator/
I had to use this because `data_dict.get('accepted_dataset_id', None)` would return an empty string if the key was present with no value. Where the correct value should be None.
It is pretty much a shorthand equivalent of `data_req.accepted_dataset_id = data_dict.get('accepted_dataset_id') if data_dict.get('accepted_dataset_id') and data_dict.get('accepted_dataset_id') != '' else None`

duttonw replied:
cool, handing empty string.

ThrawnCA commented on ckanext/datarequests/controllers/ui_controller.py:
```diff
@@ -305,6 +305,10 @@ def _return_page(errors={}, errors_summary={}):
                 c.datasets.append({'name': dataset.get('name'), 'title': dataset.get('title')})
 
+            if tk.h.closing_circumstances_enabled:
+                # This is required so the form can set the currently selected close_curcumstance option in the select dropdown
```
Misspelled "circumstance"

ThrawnCA commented on ckanext/datarequests/db.py:
```diff
@@ -164,14 +174,31 @@ def get_datarequest_followers_number(cls, **kw):
+            log.info("DataRequests-UpdateDB: 'close_circumstance' field does not exist, adding...")
+            DDL('ALTER TABLE "datarequests" ADD COLUMN "close_circumstance" varchar(100) NULL').execute(model.Session.get_bind())
+
+        if 'datarequests' in meta.tables and 'approx_publishing_date' not in meta.tables['datarequests'].columns:
```
The first part of this condition could be combined with the previous condition:

    if 'datarequests' in meta.tables:
        if 'close_circumstance' not in meta.tables['datarequests'].columns:
            ...
        if 'approx_publishing_date' not in meta.tables['datarequests'].columns:
            ...

ThrawnCA commented on test/features/datarequest_circumstances.feature:
```diff
@@ -0,0 +1,167 @@
+        | DataRequestOrgAdmin   |
+
+
+    Scenario Outline: Data request creator, Sysadmin and Admin users, when I select vlosing circumstance with condition 'To be released as open data at a later date', approx publishing date is required
```
Misspelled "closing"

duttonw commented on .docker/scripts/create-test-data.sh:
```diff
@@ -14,6 +14,21 @@ CKAN_INI_FILE=/app/ckan/default/production.ini
+#
+echo "Adding ckan.datarequests.closing_circumstances:"
+
+curl -L -s --header "Authorization: ${API_KEY}" --header "Content-Type: application/json" \
```
with our deployment to higher environments, how is this being injected. via opsworks init script or manually which we need to remember doing?

MarkCalvert replied:
I cannot recall how we did this last time but I feel like we just provided the config value and you manually did the update via the UI after deployment?
You only need to add this config value to each environment once the first time these changes are deployed, unless you drop the DB.
Or you can copy the above section from 'create-test-data.sh' and run the script on each deployment with the risk of blowing away any manual changes made to this config value from UI.
Here is the config value for 'Data request closing circumstances':
```
Released as open data|nominate_dataset
Open dataset already exists|nominate_dataset
Partially released|nominate_dataset
To be released as open data at a later date|nominate_approximate_date
Data openly available elsewhere
Not suitable for release as open data
Requested data not available/cannot be compiled
Requestor initiated closure
```

duttonw commented on ckanext/datarequests/controllers/ui_controller.py:
```diff
@@ -263,7 +263,7 @@ def delete(self, id):
             tk.abort(404, tk._('Data Request %s not found') % id)
         except tk.NotAuthorized as e:
-            log.warn(e) 
+            log.warn(e)
```
how big is the exceptions, we might want a short string instead to be logged if its a full stack trace.

MarkCalvert replied:
There is no stack trace in a NotAuthorized exception(https://github.com/ckan/ckan/blob/master/ckan/logic/__init__.py#L55), just a message (https://github.com/ckan/ckan/blob/master/ckan/logic/__init__.py#L31) that can be defined in the auth function. But after checking the auth functions from the original programmer, he did not include any messages so the log which comes back is only `ckan_1 | 17:07:40,863 WARNI [ckanext.datarequests.controllers.ui_controller]`

duttonw commented on ckanext/datarequests/db.py:
```diff
@@ -164,14 +174,31 @@ def get_datarequest_followers_number(cls, **kw):
+    if closing_circumstances_enabled:
+        if 'datarequests' in meta.tables and 'close_circumstance' not in meta.tables['datarequests'].columns:
+            log.info("DataRequests-UpdateDB: 'close_circumstance' field does not exist, adding...")
+            DDL('ALTER TABLE "datarequests" ADD COLUMN "close_circumstance" varchar(100) NULL').execute(model.Session.get_bind())
```
we have a problem, in constants we have ```CLOSE_CIRCUMSTANCE_MAX_LENGTH = 255``` but we set database table to only hold 100.

duttonw commented on ckanext/datarequests/tests/test_actions.py:
```diff
@@ -36,6 +36,7 @@ def setUp(self):
         actions.USERS_CACHE = {}
         actions.tk.ObjectNotFound = self._tk.ObjectNotFound
         actions.tk.ValidationError = self._tk.ValidationError
+        actions.tk.h.closing_circumstances_enabled = False
```
do we have a unit test where this is True?

MarkCalvert replied:
No unfortunately, I only had enough time to create BDD tests which were timeboxed to 4 hours https://salsadigital.atlassian.net/browse/DQL2-40.
We can create another story for unit tests as it probably will need to be done during the open-source giveback story https://salsadigital.atlassian.net/browse/DQL2-48

duttonw reviewed (changes requested):
ddl 100 vs constant 255 needs to be corrected, also other plugin has just been merged so requirements-dev can now be updated also.

Commit:
[DQL-33]
- Added constant 'CLOSE_CIRCUMSTANCE_MAX_LENGTH' for 'close_circumstance' varchar size
- Fixed spelling mistakes

[DQl-40]
- Fixed spelling mistakes

duttonw commented on ckanext/datarequests/db.py:
```diff
@@ -195,10 +195,11 @@ def update_db(model):
-            log.info("DataRequests-UpdateDB: 'approx_publishing_date' field does not exist, adding...")
-            DDL('ALTER TABLE "datarequests" ADD COLUMN "approx_publishing_date" timestamp NULL').execute(model.Session.get_bind())
+        if 'datarequests' in meta.tables:
+            if 'close_circumstance' not in meta.tables['datarequests'].columns:
```
@ThrawnCA  did we already run the old code in dev already?  overall this would be ok but if we have already updated to include the coloumn at 100char before we moved to the new code below, we will never get the increase

MarkCalvert replied:
@duttonw I had the same issue in my local dev environment so had to drop the columns with this command `ALTER TABLE "public"."datarequests" DROP COLUMN "close_circumstance";
ALTER TABLE "public"."datarequests" DROP COLUMN "approx_publishing_date";`
Then restart the CKAN serve command, browse to the datarequest page and it should run the new updated code above.

duttonw replied:
@MarkCalvert I don't believe this is good enough as we really don't want to be running db commands manually. The application itself should handle this, i.e. increase the size but not decrease if a rollback occurs. Its already creating the new column it just needs to do the test to see if its long enough.

duttonw commented on ckanext/datarequests/db.py:
```diff
@@ -195,10 +195,11 @@ def update_db(model):
+        if 'datarequests' in meta.tables:
+            if 'close_circumstance' not in meta.tables['datarequests'].columns:
+                log.info("DataRequests-UpdateDB: 'close_circumstance' field does not exist, adding...")
+                DDL('ALTER TABLE "datarequests" ADD COLUMN "close_circumstance" varchar({0}) NULL'.format(constants.CLOSE_CIRCUMSTANCE_MAX_LENGTH)).execute(model.Session.get_bind())
```
very nice

duttonw commented on requirements-dev.txt:
```diff
@@ -8,7 +8,7 @@ nose_parameterized==0.3.3
 lxml==4.5.0
 splinter>=0.13.0
--e git+https://github.com/salsadigitalauorg/ckanext-data-qld@feature/Enhancements_to_the_data_request_functionality#egg=ckanext-data_qld
+-e git+https://github.com/qld-gov-au/ckanext-data-qld@develop#egg=ckanext-data_qld
```
👍

Commit:
[DQL2-31]
- Updated labels from 'Approx Publishing Date' to 'Approximate Publishing Date'

[DQL2-40]
- Updated BDD tests for new label text 'Approximate Publishing Date'
